third_party/amd/lib/TritonAMDGPUToLLVM/TargetInfo.cpp

```diff
@@ -200,15 +200,19 @@ bool TargetInfo::warpReduce(RewriterBase &rewriter, Location loc,
                             unsigned numLaneToReduce,
                             unsigned interleave) const {
   auto b = TritonLLVMOpBuilder(loc, rewriter);
-  if (numLaneToReduce != 64)
-    return false;
 
-  if (!llvm::is_contained(
-          {ISAFamily::CDNA2, ISAFamily::CDNA3, ISAFamily::CDNA4},
-          getISAFamily())) {
+  if (!(numLaneToReduce == 64 &&
+            llvm::is_contained(
+                {ISAFamily::CDNA2, ISAFamily::CDNA3, ISAFamily::CDNA4},
+                getISAFamily()) ||
+        numLaneToReduce == 32 &&
+            llvm::is_contained({ISAFamily::RDNA3}, getISAFamily()))) {
     return false;
   }
 
+  bool isCDNA = llvm::is_contained(
+      {ISAFamily::CDNA2, ISAFamily::CDNA3, ISAFamily::CDNA4}, getISAFamily());
```

Comment:
You can add some helper methods in `TargetInfo`, `getWarpSize()`, `isCDNA()` and `isRDNA()`, to help simplify the logic here to something like

```c++
if (numLaneToReduce != getWarpSize()) return false;
if (isCDNA() && getISAFamily() == CDNA1) return false;
if (isRDNA() && getISAFamily() != RDNA4) return false;
```